src/common/mem.c

```diff
@@ -37,52 +37,72 @@
 #define TXD_CB_NUM		(1 << 5)	/* 32 callbacks */
 #define TXD_CB_MAX		(1 << 20)	/* 1 million callbacks */
 
-struct umem_tx_stage_item {
-	int		 txi_magic;
-	umem_tx_cb_t	 txi_fn;
-	void		*txi_data;
-};
+/** Convert an offset to an id.   No invalid flags will be maintained
+ *  in the conversion.
+ *
+ *  \param	umm[IN]		The umem pool instance
+ *  \param	umoff[in]	The offset to convert
+ *
+ *  \return	The oid.
+ */
+static inline PMEMoid
+umem_off2id(const struct umem_instance *umm, umem_off_t umoff)
+{
+	PMEMoid	oid;
 
-/** persistent memory operations (depends on pmdk) */
+	if (UMOFF_IS_NULL(umoff))
+		return OID_NULL;
 
-static umem_id_t
-pmem_id(const struct umem_instance *umm, void *addr)
-{
-	return pmemobj_oid(addr);
-}
+	oid.pool_uuid_lo = umm->umm_pool_uuid_lo;
+	oid.off = umem_off2offset(umoff);
 
-static void *
-pmem_addr(const struct umem_instance *umm, umem_id_t ummid)
-{
-	return pmemobj_direct(ummid);
+	return oid;
 }
 
-static bool
-pmem_equal(struct umem_instance *umm, umem_id_t ummid1, umem_id_t ummid2)
+/** Convert an id to an offset.
+ *
+ *  \param	umm[IN]		The umem pool instance
+ *  \param	oid[in]		The oid to convert
+ *
+ *  \return	The offset in the PMEM pool.
+ */
+static inline umem_off_t
+umem_id2off(const struct umem_instance *umm, PMEMoid oid)
 {
-	return OID_EQUALS(ummid1, ummid2);
+	if (OID_IS_NULL(oid))
+		return UMOFF_NULL;
+
+	return oid.off;
 }
 
+struct umem_tx_stage_item {
+	int		 txi_magic;
```

Comment:
Probably obvious to most, but what is "magic"?

Reply:
@ryon-jensen It's used for sanity check that the item is a valid one.  Kind of like a weak type check.  It's set when adding a callback and checked when using it with D_ASSERT